marconi/src/Marconi/Index/MintBurn.hs

```diff
@@ -0,0 +1,310 @@
+fromMaryPolicyID (LM.PolicyID sh) = C.PolicyId (C.fromShelleyScriptHash sh) -- from cardano-api:src/Cardano/Api/Value.hs
+
+fromMaryAssetName :: LM.AssetName -> C.AssetName
+fromMaryAssetName (LM.AssetName n) = C.AssetName $ Short.fromShort n -- from cardano-api:src/Cardano/Api/Value.hs
+
+fromAlonzoData :: LA.Data ledgerera -> C.ScriptData
+fromAlonzoData = C.fromPlutusData . LA.getPlutusData -- from cardano-api:src/Cardano/Api/ScriptData.hs
+
+-- * Sqlite
+
+sqliteInit :: SQL.Connection -> IO ()
+sqliteInit c = liftIO $ do
+  SQL.execute_ c
+    " CREATE TABLE IF NOT EXISTS        \
+    \   minting_policy_events           \
+    \   ( slotNo          INT NOT NULL  \
+    \   , blockHeaderHash INT NOT NULL  \
+    \   , txId            BLOB NOT NULL \
+    \   , policyId        BLOB NOT NULL \
+    \   , assetName       TEXT NOT NULL \
+    \   , quantity        INT NOT NULL  \
+    \   , redeemerIx      INT NOT NULL  \
+    \   , redeemerData    BLOB NOT NULL)"
+  SQL.execute_ c
+    " CREATE INDEX IF NOT EXISTS               \
+    \    minting_policy_events__txId_policyId  \
+    \ ON minting_policy_events (txId, policyId)"
+
+sqliteInsert :: SQL.Connection -> [TxMintEvent] -> IO ()
+sqliteInsert c es = SQL.executeMany c template $ toRows =<< toList es
+  where
+    template =
+      "INSERT INTO minting_policy_events \
+      \ ( slotNo, blockHeaderHash, txId  \
+      \ , policyId, assetName, quantity  \
+      \ , redeemerIx, redeemerData )     \
+      \ VALUES (?, ?, ?, ?, ?, ?, ?, ?)  "
+
+    toRows :: TxMintEvent -> [[SQL.SQLData]]
+    toRows e = do
+      (txId, txMintAssets) <- NE.toList $ txMintEventTxAssets e
+      mintAsset <- NE.toList txMintAssets
+      pure
+        [ SQL.toField $ txMintEventSlotNo e
+        , SQL.toField $ txMintEventBlockHeaderHash e
+        , SQL.toField txId
+        , SQL.toField $ mintAssetPolicyId mintAsset
+        , SQL.toField $ mintAssetAssetName mintAsset
+        , SQL.toField $ mintAssetQuantity mintAsset
+        , SQL.toField $ mintAssetRedeemerIdx mintAsset
+        , SQL.toField $ mintAssetRedeemerData mintAsset
+        ]
+
+type Row = (C.SlotNo, C.Hash C.BlockHeader, C.TxId, C.PolicyId, C.AssetName, C.Quantity, Word64, C.ScriptData)
+
+-- | Input rows must be sorted by C.SlotNo.
+fromRows :: [Row] -> [TxMintEvent]
+fromRows rows =  do
+  rs@(r :| _) <- NE.groupBy ((==) `on` slotNo) rows -- group by SlotNo
+  pure $ TxMintEvent (slotNo r) (hash r) $ do
+    rs' <- NE.groupBy1 ((==) `on` txId) rs -- group by TxId
+    pure (txId r, rowToMintAsset <$> rs')
+  where
+    slotNo = view L._1 :: Row -> C.SlotNo
+    hash = view L._2 :: Row -> C.Hash C.BlockHeader
+    txId = view L._3 :: Row -> C.TxId
+    rowToMintAsset :: Row -> MintAsset
+    rowToMintAsset row = MintAsset (row^.L._4) (row^.L._5) (row^.L._6) (row^.L._7) (row^.L._8)
+
+sqliteSelectByTxIdPolicyId :: SQL.Connection -> (SQL.Query, [NamedParam]) -> C.TxId -> C.PolicyId -> IO [TxMintEvent]
```

Comment:
:heart:

Reply:
I agree -- it saved me copy-pasting of the same sql query multiple times. I think it was also you who started using the named parameters, I seem to remember them from a pull request but only now got around using them.